README.md
# ib-ruby

Ruby Implementation of the Interactive Brokers Trader Workstation (TWS) API v.965-967.
## Gateway-Branch, Environment: Ruby 2.2, ActiveModel,  Rspec3/Guard-Testsuite

The hole TWS-Environment is accessible through Ruby-Objects.

IB::Gateway is the root. It manages a list of TWS-Users. 
An Advisor, who manages a list of contract-queries
and ActiveAccounts, where AccountValues, PortfolioValues and Orders are linked.

Whenever queries are send to the TWS, the response is stored in the object-tree
and can then read out with standard Array-Methods.

Thus ib-ruby supports the following workflow

* Application sends Request 
* IB::Gateway transmits to the TWS
* TWS-Response is stored in Object-Tree
* Application gets Response
* Application reads the evaluated response from Object-Tree

Time-critical operations are encapsulated in IB::Connection, which itself is
managed by IB::Gateway. IB::Gateway takes care of interrupted connections to the TWS
and tolerates the daily reset of the TWS, and thus enables a 24/7-operation-mode.

However, ib-ruby offers a simple translation of ruby-queries to tws-socket-codes and
offers the pure TWS-response as well. The usage of the object-tree is optional.
Any code for previous versions of the program should work.

For more details refer to the [introduction](intro.md)
and for programming hints the [integration](integration.md) section.

### Changes from the stable branch


* Only ActiveModel-Support. 
* Alert-Messages are handled by IB::Alerts
* IB::Stock, IB::Future, IB::Forex are derived from IB::Contract
* IB::Account model added, where contracts, orders, positions and AccountValues are present
* IB::Gateway builds an object-orientated representation of Accounts with pending and completed 
Orders, Positions and Contracts. A thread-safe access to objects which are updated concurrently 
by the TWS is realized.

An Example
``` ruby
    gw = IB::Gateway.new get_account_data:true  # connects to the TWS by default
    accounts = gw.active_accounts
    accounts.each do |account|
     puts account.simple_account_data_scan('AccountCode')
     puts account.simple_account_data_scan('TotalCashValue')
     puts account.contracts.map &:to_human 
     puts account.portfolio_values &:to_human
    end
    gw.disconnect
 ``` 
 leads to
 ```
 <AccountCode=DU167348 >
 <TotalCashValue=601740.49 EUR>
 <TotalCashValue-C=29290.41 EUR>
 <TotalCashValue-S=572450.08 EUR>
 <Stock: BLUE EUR>
 <Stock: CBA AUD>
 <Stock: CIEN USD>
 <PortfolioValue: <Stock: BLUE EUR> (720): Market 25.3299999 price 18237.6 value; PnL: 1934.31 unrealized, 0.0 realized;>
 <PortfolioValue: <Stock: CBA AUD> (1004): Market 83.1100006 price 83442.44 value; PnL: 3761.55 unrealized, 0.0 realized;
 ```
* To Query the TWS manualy, the IB::Connection-Object is always available via IB::Gateway.tws, eg.
```ruby
   IB::Gateway.tws.send_message(...)
   IB::Gateway.tws.subscribe(...)
```
The previous way to access the TWS by initializing IB::Connection is still supported. 

IB::Gateway.tws replaces IB::Connection.current 






Copyright (C) 2006-2015 Paul Legato, Wes Devauld, Ar Vicco and Hartmut Bischoff.

https://github.com/ib-ruby/ib-ruby

__WARNING:__ This software is provided __AS-IS__ with __NO WARRANTY__, express or
implied. Your use of this software is at your own risk. It may contain any number
of bugs, known or unknown, which might cause you to lose money if you use it.
You've been warned.

This code is not sanctioned or supported by Interactive Brokers.

##ANNOUNCE:
Checkout Branch »Gateway«  for support of FA-(aka Friends & Family)-Accounts.
Its a release-candidate for Version 0.9.3

## SUMMARY:

This is a pure Ruby implementation of Interactive Brokers API. It is NOT a wrapper
for a Java or C++ API, but rather uses socket API directly. So it does not have any
dependencies other than TWS/Gateway itself.

Why Ruby? Many people are put off by the amount of boilerplate code/plumbing required
by Java, ActiveX or C++ API to do even the simplest of things, like getting account
data and placing/monitoring orders. This library intends to keep all the fluff away
and let you focus on writing your business logics, rather than useless boilerplate.

No more endless definitions of obligatory methods you'd never need, no spaghetti code
to divide your execution flow between multiple callbacks and interfaces.

Instead, a very simple paradigm is offered: your code interacts with the server
(TWS or Gateway) via exchange of messages. You subscribe to the server messages
that you're interested in, and send messages to server that request specific data
from it. You wait for specific messages being received, or other conditions you
define. The execution flow is under your control, rather than delegated somewhere.

Using this clear paradigm, you can hack together a simple automation of your
daily TWS-related routine in just a couple of minutes. Alternatively, you can
create a mechanical trading system with complex order processing logics, that
contains 1/10th of code and is 500% more maintaineable than it is possible with
other API implementations. The choice is yours.

## INSTALLATION:

### From Source

    $ git clone https://github.com/ib-ruby/ib-ruby
    $ cd ib-ruby; rake gem:install

## PREREQUISITES:

1. Install Interactive Brokers connectivity software: either
   [TWS](http://www.interactivebrokers.com/en/p.php?f=tws) or
   [Gateway](http://www.interactivebrokers.com/en/p.php?f=programInterface&ib_entity=llc)

2. Configure the software to allow API connections from the computer you plan to run
   ib-ruby on, which is typically localhost (127.0.0.1) if you're running ib-ruby on
   the same machine as TWS/Gateway. [Here](http://www.youtube.com/watch?v=53tmypRq5wI)
   you can see how this is done for TWS.

3. Make sure sure your ib-ruby gem version is compatible with your software version.
   As a rule of thumb, most recent ib-ruby gem only supports latest versions of
   TWS/Gateway API. Older versions of API are supported by previous gem versions:

    | ib-ruby gem | TWS version | API version  |
    |:------------|------------:|:------------:|
    | 0.5.21      |    918-920  |    965       |
    | 0.6.1       |    921-923  |    966       |
    | 0.7.1       |    924-925  |    966       |
    | 0.8.1       |    926-930  |    967 beta  |
    | 0.9.0+      |    931-932  |    967 final |

4. Start Interactive Broker's Trader Work Station or Gateway before your code
   attempts to connect to it. Note that TWS and Gateway listen to different ports,
   this library assumes connection to Gateway on the same machine (localhost:4001)
   by default, this can be changed via :host and :port options given to IB::Connection.new.

## SYNOPSIS:

This is an example of your script that requests and prints out account data, then
places limit order to buy 100 lots of WFC and waits for execution. All in about ten
lines of code - and without sacrificing code readability or flexibility.
``` ruby
    require 'ib'

    ib = IB::Connection.new :port => 7496
    ib.subscribe(:Alert, :AccountValue) { |msg| puts msg.to_human }
    ib.send_message :RequestAccountData
    ib.wait_for :AccountDownloadEnd

    ib.subscribe(:OpenOrder) { |msg| puts "Placed: #{msg.order}!" }
    ib.subscribe(:ExecutionData) { |msg| puts "Filled: #{msg.execution}!" }
    contract = IB::Contract.new :symbol => 'WFC', :exchange => 'NYSE',
                                :currency => 'USD', :sec_type => :stock
    buy_order = IB::Order.new :total_quantity => 100, :limit_price => 21.00,
                                :action => :buy, :order_type => :limit
    ib.place_order buy_order, contract
    ib.wait_for :ExecutionData
```
Your code interacts with TWS via exchange of messages. Messages that you send to
TWS are called 'Outgoing', messages your code receives from TWS - 'Incoming'.

First, you need to subscribe to incoming message types you're interested in
using `Connection#subscribe`. The code block (or proc) given to `#subscribe`
will be executed when an incoming message of the this type is received from TWS,
with the received message as its argument.

Then, you request specific data from TWS using `Connection#send_message` or place
your order using `Connection#place_order`. TWS will respond with messages that you
should have subscribed for, and these messages will be processed in a code block
given to `#subscribe`.

In order to give TWS time to respond, you either run a message processing loop or
just wait until Connection receives the messages type you requested.

See `lib/ib/messages` for a full list of supported incoming/outgoing messages
and their attributes. The original TWS docs and code samples can also be found
in `misc` directory.

Sample scripts in `example` directory demonstrate common ib-ruby use cases. Examples
show you how to access account info, print real time quotes, retrieve historic or
fundamental data, request options calculations, place, list, and cancel orders.
You may also want to look into `spec/integration` directory for more scenarios,
use cases and examples of handling IB messages.



## DB BACKEND:

If you want to take advantage of data persistance layer ActiveRecord ORM, you have to set up the database
(SQLite recommended for simplicity) and run migrations located at gems 'db/migrate' folder.

You further need to:
``` ruby
    require 'ib/db'
    IB::DB.connect :adapter => 'sqlite3', :database => 'db/test.sqlite3'
    require 'ib'
```
Only require 'ib' AFTER you've connected to DB, otherwise your Models will not
inherit from ActiveRecord::Base and won't be persistent. 

Now, all your IB Models are just ActiveRecords and you can save them to the DB.

## RUNNING TESTS:

The gem comes with a spec suit that may be used to test ib-ruby compatibility with your
specific TWS/Gateway installation. Please read 'spec/Readme.md' for more details about
running specs.


## CONTRIBUTING:

If you want to contribute to ib-ruby development:

1. Make a fresh fork of ib-ruby (Fork button on top of Github GUI)
2. Clone your fork locally (git clone /your fork private URL/)
3. Add main ib-ruby repo as upstream (git remote add upstream git://github.com/ib-ruby/ib-ruby.git)
4. Create your feature branch (git checkout -b my-new-feature)
5. Modify code as you see fit
6. Commit your changes (git commit -am 'Added some feature')
7. Pull in latest upstream changes (git fetch upstream -v; git merge upstream/master)
8. Push to the branch (git push origin my-new-feature)
9. Go to your Github fork and create new Pull Request via Github GUI

... then proceed from step 5 for more code modifications...

## LICENSE:

This software is available under the LGPL.

This library is free software; you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as
published by the Free Software Foundation; either version 2.1 of the
License, or (at your option) any later version.

This library is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the file LICENSE
for full licensing details of GNU Lesser General Public License.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
02110-1301 USA



HISTORY
== 0.1 / 2006-11-12

* Initial version - Paul Legato

== 0.2 / 2007-02-14

* Alpha Release - Paul Legato

== 0.3 / 2009-06-18

* Gem packaging

== 0.4 / 2011-09-04

* Repackaging with Bundler and Gemfile

== 0.4.3 / 2011-09-05

* Old tests converted into specs

== 0.4.4 / 2011-09-05

* IB references and Java examples added

== 0.4.6 / 2011-09-07

* CSV Downloader Java examples added

== 0.4.7 / 2011-09-12

* Datatypes turned into Models

== 0.4.14 / 2011-09-18

* DOM script added

== 0.4.19 / 2011-09-19

* All IB messages implemented

== 0.4.20 / 2011-09-19

* IB::Connection#subscribe now accepts Symbols with message names

== 0.4.22 / 2011-09-20

* Fixed minor buglets

== 0.5.0 / 2011-10-29

* Rake versioning bug resolved

== 0.5.1 / 2011-10-29

* Connection specs extended

== 0.5.2 / 2011-10-30

* Add nonblocking Connection#process_messages API

== 0.5.4 / 2011-12-16

* Placing Version source file into its proper place

== 0.5.5 / 2011-12-16

* ComboLeg serialization signature changed

== 0.5.6 / 2011-12-16

* Models::ComboLeg given a source file of its own

== 0.5.7 / 2011-12-16

* Option subclass of Contract added

== 0.5.8 / 2011-12-16

* Critical bugs in Order-related messages fixed

== 0.5.9 / 2011-12-16

* Sample scripts to place and cancel orders added

== 0.5.10 / 2011-12-19

* Bag Contract subclass added

== 0.5.11 / 2011-12-20

* Release Connection.current API

== 0.5.12 / 2011-12-21

* Logging support hacked back in

== 0.5.13 / 2012-01-03

* Critical bug in Connection#place_order fixed

== 0.5.14 / 2012-01-28

* Bump after Rebundling

== 0.5.15 / 2012-02-01

* Docs improvement

== 0.5.16 / 2012-02-08

* Message specs refactored

== 0.5.17 / 2012-02-12

* Connection#send_message bugfix

== 0.5.18 / 2012-02-15

* Bugfix release

== 0.5.19 / 2012-02-23

* Contract

== 0.5.21 / 2012-03-06

* Fully tested with integration spec suite

== 0.6.0 / 2012-03-09

* Connection wait_for/received/received? API added

== 0.6.1 / 2012-03-14

* Version bump

== 0.7.0 / 2012-03-22

* Support for API v.967

== 0.7.1 / 2012-03-22

* Starting preparation for ActiveModel compatibility

== 0.7.2 / 2012-03-23

* Better spec documentation

== 0.7.3 / 2012-03-23

* Connection API extended

== 0.7.4 / 2012-04-04

* Order equality bug fixed

== 0.7.5 / 2012-04-16

* Optional DB backend for Models /ActiveRecord/

== 0.7.6 / 2012-04-18

* Bugfix for FundamentalData

== 0.7.7 / 2012-04-20

* Model validations and DB backend

== 0.7.8 / 2012-04-23

* Models are now fully DB-backed

== 0.7.9 / 2012-04-24

* Order and OrderState API extended

== 0.7.10 / 2012-04-25

* Small fixes for DB backend and serialization

== 0.7.11 / 2012-04-25

* DB schema extended

== 0.7.12 / 2012-04-27

* ComboLeg is a join Model

== 0.8.0 / 2012-06-07

* TWS 926 support

== 0.8.1 / 2012-06-13

* Combo routing parameters and algo parameters supported

== 0.8.4 / 2012-07-18

* Working Rails Engine integration

== 0.8.5 / 2012-08-11

* Complete Rails Engine, TWS 929 support

== 0.9.0 / 2012-10-31

* Support for API 967 final

== 0.9.3 / 2015

* Gateway Overlay which manages Accounts


lib/ib/constants.rb
module IB
  ### Widely used TWS constants:

  EOL = "\0"

  # Enumeration of bar size types for convenience.
  # Bar sizes less than 30 seconds do not work for some securities.
  BAR_SIZES = {'1 sec' => :sec1,
               '5 secs' => :sec5,
               '15 secs' =>:sec15,
               '30 secs' =>:sec30,
               '1 min' => :min1,
               '2 mins' => :min2,
               '3 mins' => :min3,
               '5 mins' => :min5,
               '15 mins' =>:min15,
               '30 mins' =>:min30,
               '1 hour' =>:hour1,
	       '2 hours' => :hour2,
	       '4 hours' => :hour4,
               '1 day' => :day1,
               '1 week' => :week1
  }.freeze

  # Enumeration of data types.
  # Determines the nature of data being extracted. Valid values:
  DATA_TYPES = {'TRADES' => :trades,
                'MIDPOINT' => :midpoint,
                'BID' => :bid,
                'ASK' => :ask,
                'BID_ASK' => :bid_ask,
                'HISTORICAL_VOLATILITY' => :historical_volatility,
                'OPTION_IMPLIED_VOLATILITY' => :option_implied_volatility,
                'OPTION_VOLUME' => :option_volume,
                'OPTION_OPEN_INTEREST' => :option_open_interest
  }.freeze

  ### These values are typically received from TWS in incoming messages

  # Tick types as received in TickPrice and TickSize messages (enumeration)
  TICK_TYPES = {
      # int id => :Description #  Corresponding API Event/Function/Method
      0 => :bid_size, #               tickSize()
      1 => :bid_price, #              tickPrice()
      2 => :ask_price, #              tickPrice()
      3 => :ask_size, #               tickSize()
      4 => :last_price, #             tickPrice()
      5 => :last_size, #              tickSize()
      6 => :high, #                   tickPrice()
      7 => :low, #                    tickPrice()
      8 => :volume, #                 tickSize()
      9 => :close_price, #            tickPrice()
      10 => :bid_option, #            tickOptionComputation() See Note 1 below
      11 => :ask_option, #            tickOptionComputation() See => :Note 1 below
      12 => :last_option, #           tickOptionComputation()  See Note 1 below
      13 => :model_option, #          tickOptionComputation() See Note 1 below
      14 => :open_tick, #             tickPrice()
      15 => :low_13_week, #           tickPrice()
      16 => :high_13_week, #          tickPrice()
      17 => :low_26_week, #           tickPrice()
      18 => :high_26_week, #          tickPrice()
      19 => :low_52_week, #           tickPrice()
      20 => :high_52_week, #          tickPrice()
      21 => :avg_volume, #            tickSize()
      22 => :open_interest, #         tickSize()
      23 => :option_historical_vol, # tickGeneric()
      24 => :option_implied_vol, #    tickGeneric()
      25 => :option_bid_exch, #   not USED
      26 => :option_ask_exch, #   not USED
      27 => :option_call_open_interest, # tickSize()
      28 => :option_put_open_interest, #  tickSize()
      29 => :option_call_volume, #        tickSize()
      30 => :option_put_volume, #         tickSize()
      31 => :index_future_premium, #      tickGeneric()
      32 => :bid_exch, #                  tickString()
      33 => :ask_exch, #                  tickString()
      34 => :auction_volume, #    not USED
      35 => :auction_price, #     not USED
      36 => :auction_imbalance, # not USED
      37 => :mark_price, #              tickPrice()
      38 => :bid_efp_computation, #     tickEFP()
      39 => :ask_efp_computation, #     tickEFP()
      40 => :last_efp_computation, #    tickEFP()
      41 => :open_efp_computation, #    tickEFP()
      42 => :high_efp_computation, #    tickEFP()
      43 => :low_efp_computation, #     tickEFP()
      44 => :close_efp_computation, #   tickEFP()
      45 => :last_timestamp, #          tickString()
      46 => :shortable, #               tickGeneric()
      47 => :fundamental_ratios, #      tickString()
      48 => :rt_volume, #               tickGeneric()
      49 => :halted, #           see Note 2 below.
      50 => :bid_yield, #               tickPrice() See Note 3 below
      51 => :ask_yield, #               tickPrice() See Note 3 below
      52 => :last_yield, #              tickPrice() See Note 3 below
      53 => :cust_option_computation, # tickOptionComputation()
      54 => :trade_count, #             tickGeneric()
      55 => :trade_rate, #              tickGeneric()
      56 => :volume_rate, #             tickGeneric()
      57 => :last_rth_trade, #            ?
      #   Note 1: Tick types BID_OPTION, ASK_OPTION, LAST_OPTION, and MODEL_OPTION return
      #           all Greeks (delta, gamma, vega, theta), the underlying price and the
      #           stock and option reference price when requested.
      #           MODEL_OPTION also returns model implied volatility.
      #   Note 2: When trading is halted for a contract, TWS receives a special tick:
      #           haltedLast=1. When trading is resumed, TWS receives haltedLast=0.
      #           A tick type, HALTED, tick ID= 49, is now available in regular market
      #           data via the API to indicate this halted state. Possible values for
      #           this new tick type are: 0 = Not halted, 1 = Halted.
      #   Note 3: Applies to bond contracts only.
  }

  # Financial Advisor types (FaMsgTypeName)
  FA_TYPES = {
      1 => :groups,
      2 => :profiles,
      3 => :aliases}.freeze

  # Received in new MarketDataType (58 incoming) message
  MARKET_DATA_TYPES = {
      0 => :unknown,
      1 => :real_time,
      2 => :frozen,
  }

  # Market depth messages contain these "operation" codes to tell you what to do with the data.
  # See also http://www.interactivebrokers.com/php/apiUsersGuide/apiguide/java/updatemktdepth.htm
  MARKET_DEPTH_OPERATIONS = {
      0 => :insert, # New order, insert into the row identified by :position
      1 => :update, # Update the existing order at the row identified by :position
      2 => :delete # Delete the existing order at the row identified by :position
  }.freeze

  MARKET_DEPTH_SIDES = {
      0 => :ask,
      1 => :bid
  }.freeze

  ORDER_TYPES =
      {'LMT' => :limit, #                  Limit Order
       'LIT' => :limit_if_touched, #       Limit if Touched
       'LOC' => :limit_on_close, #         Limit-on-Close      LMTCLS ?
       'LOO' => :limit_on_open, #          Limit-on-Open
       'MKT' => :market, #                 Market
       'MIT' => :market_if_touched, #      Market-if-Touched
       'MOC' => :market_on_close, #        Market-on-Close     MKTCLSL ?
       'MOO' => :market_on_open, #         Market-on-Open
       'MTL' => :market_to_limit, #        Market-to-Limit
       'MKTPRT' => :market_protected, #   Market with Protection
       'QUOTE' => :request_for_quote, #    Request for Quote
       'STP' => :stop, #                   Stop
       'STPLMT' => :stop_limit, #         Stop Limit
       'TRAIL' => :trailing_stop, #        Trailing Stop
       'TRAIL LIMIT' => :trailing_limit, # Trailing Stop Limit
       'TRAIL LIT' => :trailing_limit_if_touched, #  Trailing Limit if Touched
       'TRAIL MIT' => :trailing_market_if_touched, # Trailing Market If Touched
       'PEG MKT' => :pegged_to_market, #   Pegged-to-Market
       'REL' => :relative, #               Relative
       'BOX TOP' => :box_top, #            Box Top
       'PEG MID' => :pegged_to_midpoint, # Pegged-to-Midpoint
       'VWAP' => :vwap, #                  VWAP-Guaranteed
       'OCA' => :one_cancels_all, #        One-Cancels-All
       'VOL' => :volatility, #             Volatility
       'SCALE' => :scale, #                Scale
       'NONE' => :none, # Used to indicate no hedge in :delta_neutral_order_type
       'None' => :none, # Used to indicate no hedge in :delta_neutral_order_type
      }.freeze

  # Valid security types (sec_type attribute of IB::Contract)
  SECURITY_TYPES =
      {'STK' => :stock,
       'OPT' => :option,
       'FUT' => :future,
       'IND' => :index,
       'FOP' => :futures_option,
       'CASH' => :forex,
       'BOND' => :bond,
       'WAR' => :warrant,
       'FUND' => :fund, # ETF?
       'BAG' => :bag}.freeze

  # Obtain symbolic value from given property code:
  # VALUES[:side]['B'] -> :buy
  VALUES = {
      :sec_type => SECURITY_TYPES,
      :order_type => ORDER_TYPES,
      :delta_neutral_order_type => ORDER_TYPES,

      :origin => {0 => :customer, 1 => :firm},
      :volatility_type => {1 => :daily, 2 => :annual},
      :reference_price_type => {1 => :average, 2 => :bid_or_ask},

      # This property encodes differently for ComboLeg and Order objects,
      # we use ComboLeg codes and transcode for Order codes as needed
      :open_close =>
          {0 => :same, # Default for Legs, same as the parent (combo) security.
           1 => :open, #  Open. For Legs, this value is only used by institutions.
           2 => :close, # Close. For Legs, this value is only used by institutions.
           3 => :unknown}, # WTF

      :right =>
          {'' => :none, # Not an option
           'P' => :put,
           'C' => :call},

      :side => # AKA action
          {'B' => :buy, # or BOT
           'S' => :sell, # or SLD
           'T' => :short, # short
           'X' => :short_exempt # Short Sale Exempt action. This allows some orders
           # to be exempt from the SEC recent changes to Regulation SHO, which
           # eliminated the old uptick rule and replaced it with a new "circuit breaker"
           # rule, and allows some orders to be exempt from the new rule.
          },

      :short_sale_slot =>
          {0 => :default, #      The only valid option for retail customers
           1 => :broker, #       Shares are at your clearing broker, institutions
           2 => :third_party}, # Shares will be delivered from elsewhere, institutions

      :oca_type =>
          {0 => :none, # Not a member of OCA group
           1 => :cancel_with_block, # Cancel all remaining orders with block
           2 => :reduce_with_block, # Remaining orders are reduced in size with block
           3 => :reduce_no_block}, # Remaining orders are reduced in size with no block

      :auction_strategy =>
          {0 => :none, # Not a BOX order
           1 => :match,
           2 => :improvement,
           3 => :transparent},

      :trigger_method =>
          {0 => :default, # "double bid/ask" used for OTC/US options, "last" otherswise.
           1 => :double_bid_ask, # stops are triggered by 2 consecutive bid or ask prices.
           2 => :last, # stops are triggered based on the last price.
           3 => :double_last,
           4 => :bid_ask, # bid >= trigger price for buy orders, ask <= trigger for sell orders
           7 => :last_or_bid_ask, # bid OR last price >= trigger price for buy orders
           8 => :mid_point}, # midpoint >= trigger price for buy orders and the
      #      spread between the bid and ask must be less than 0.1% of the midpoint

      :hedge_type =>
          {'D' => :delta, # parent order is an option and the child order is a stock
           'B' => :beta, # offset market risk by entering into a position with
           #               another contract based on the system or user-defined beta
           'F' => :forex, # offset risk with currency different from your base currency
           'P' => :pair}, # trade a mis-valued pair of contracts and provide the
      #                     ratio between the parent and hedging child order

      :clearing_intent =>
          {'' => :none,
           'IB' => :ib,
           'AWAY' => :away,
           'PTA' => :post_trade_allocation},

      :delta_neutral_clearing_intent =>
          {'' => :none,
           'IB' => :ib,
           'AWAY' => :away,
           'PTA' => :post_trade_allocation},

      :tif =>
          {'DAY' => :day,
           'GAT' => :good_after_time,
           'GTD' => :good_till_date,
           'GTC' => :good_till_cancelled,
           'IOC' => :immediate_or_cancel},

      :rule_80a =>
          {'I' => :individual,
           'A' => :agency,
           'W' => :agent_other_member,
           'J' => :individual_ptia,
           'U' => :agency_ptia,
           'M' => :agent_other_member_ptia,
           'K' => :individual_pt,
           'Y' => :agency_pt,
           'N' => :agent_other_member_pt},

      :opt? => # TODO: unknown Order property, like OPT_BROKER_DEALER... in Order.java
          {'?' => :unknown,
           'b' => :broker_dealer,
           'c' => :customer,
           'f' => :firm,
           'm' => :isemm,
           'n' => :farmm,
           'y' => :specialist},

  }.freeze

  # Obtain property code from given symbolic value:
  # CODES[:side][:buy] -> 'B'
  CODES = Hash[VALUES.map { |property, hash| [property, hash.invert] }].freeze

  # Most common property processors
  PROPS = {:side =>
               {:set => proc { |val| # BUY(BOT)/SELL(SLD)/SSHORT/SSHORTX
                 self[:side] = case val.to_s.upcase
                                 when /SHORT.*X|\AX\z/
                                   'X'
                                 when /SHORT|\AT\z/
                                   'T'
                                 when /\AB/
                                   'B'
                                 when /\AS/
                                   'S'
                               end },
                :validate =>
                    {:format =>
                         {:with => /\Abuy\z|\Asell\z|\Ashort\z|\Ashort_exempt\z/,
                          :message => "should be buy/sell/short"}
                    }
               },

           :open_close =>
               {:set => proc { |val|
                 self[:open_close] = case val.to_s.upcase[0..0]
                                       when 'S', '0' # SAME
                                         0
                                       when 'O', '1' # OPEN
                                         1
                                       when 'C', '2' # CLOSE
                                         2
                                       when 'U', '3' # Unknown
                                         3
                                     end
               },
                :validate =>
                    {:format =>
                         {:with => /\Asame\z|\Aopen\z|\Aclose\z|\Aunknown\z/,
                          :message => "should be same/open/close/unknown"}
                    },
               }
  }.freeze

end # module IB
